// @goat-sdk/core/dist/chunk-LBC2NNGY.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _chunkUK53HMGYjs = require('./chunk-UK53HMGY.js');


var _chunk7QVYU63Ejs = require('./chunk-7QVYU63E.js');

// src/classes/WalletClientBase.ts
var _zod = require('zod');
var WalletClientBase = class {
  static {
    _chunk7QVYU63Ejs.__name.call(void 0, this, "WalletClientBase");
  }
  getCoreTools() {
    return [
      _chunkUK53HMGYjs.createTool.call(void 0, {
        name: "get_address",
        description: "Get the address of the wallet",
        parameters: _zod.z.object({})
      }, () => this.getAddress()),
      _chunkUK53HMGYjs.createTool.call(void 0, {
        name: "get_chain",
        description: "Get the chain of the wallet",
        parameters: _zod.z.object({})
      }, () => this.getChain()),
      _chunkUK53HMGYjs.createTool.call(void 0, {
        name: "get_balance",
        description: "Get the balance of the wallet",
        parameters: _zod.z.object({
          address: _zod.z.string()
        })
      }, (parameters) => this.balanceOf(parameters.address))
    ];
  }
};



exports.WalletClientBase = WalletClientBase;
